### ISSUE TITLE
Const Generic value declarations are not required to be used in struct definition

### ISSUE THREAD

**Issue by jamesmunns** [rust-lang/rust#83330]
I tried this code:

[Playground link](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=980d9b12e16ac00843b95ed79bedbdb9)

```rust
// Doesn't compile: Unused parameter
// pub struct Nothing<ANY>
// {
//     nothing: ()
// }

// Use of N not enforced
pub struct JustNum<const N: usize> {
    nothing: ()
}

// Use of N not enforced
pub struct Generic<ANY, const N: usize>
{
    any: ANY
}

// Use of N not enforced
impl<const N: usize> JustNum<{ N }> {
    fn new() -> Self {
        Self { nothing: () } 
    }
}

// Does not compile: Extra "M" 
// impl<const N: usize, const M: usize> JustNum<{ N }, { M }> {
//     fn new() -> Self {
//         Self { nothing: () } 
//     }
// }
```

I expected to see this happen: 

For the following definition:

```rust
pub struct JustNum<const N: usize> {
    nothing: ()
}
```

I expect a compiler error, due to the unused const-generic argument

Instead, this happened: 

No error was emitted

### Meta

This occurs in the current beta and nightly:

* 1.51.0-beta.8 (73f48e5f6 2021-03-18)
* 1.52.0-nightly (f5f33ec0e 2021-03-19)

I'm not sure if const generic arguments are required to be used, but I could see this as an "odd" way to tag extra parameters on without them being bound to the actual impls.

**Comment by jamesmunns** [comment]
@jonas-schievink mentions this is intended behavior due to consts not having variance ([here on twitter](https://twitter.com/sheevink/status/1373400335688040453)), if this is correct, I'm happy to close the issue.

I can't see anything "wrong" this could allow (other than acting as an associated const implicitly, I guess?), it was just a surprise for me.

**Comment by JulianKnodt** [comment]
I think maybe that's from a very compiler-oriented standpoint, for consts at least having it not be required to be instantiated allows for compile-time configuration, which for me personally is convenient, thinking of mainly linear algebra use-cases or passing compile time flags with structs. If they did have to be used, they would always have to result in some sort of `[T; N]` somewhere.

**Comment by ehuss** [comment]
It was explicitly enabled in #60058 (see #60047).  It would be nice to have an example of where that is actually useful.  Otherwise, I don't see why it shouldn't at least trigger an unused lint just like an unused function parameter does.

**Comment by Lokathor** [comment]
Sure, here's an example of a type that uses a const but doesn't store the const in a field: <https://github.com/rust-console/voladdress/blob/main/src/volblock.rs>

**Comment by jamesmunns** [comment]
Awesome, thanks for the clarification! Closing as this is intended behavior.
